infra: document instructions for yubikeys and piv (#1086)
rust-lang-nursery/rust-forge: src/infra/docs/hardware-security-keys.md

```shell

> uvx --from yubikey-manager ykman --help
> 
> # or alternatively
> 
> alias ykman="uvx --from yubikey-manager ykman"
> ykman --help

Reviewer: ````suggestion
```

or alternatively:

```
alias ykman="uvx --from yubikey-manager ykman"
ykman --help
````
So that people can copy to the clipboard one of the two commands.

Author: Done [here](https://github.com/rust-lang/rust-forge/pull/1086/changes/c5b80e4037e382143da497704f044b0da43777a7)

app.

```shell
$ ykman fido credentials list
Enter your PIN:
Credential ID  RP ID    Username  Display name
86707903...    ssh:git  openssh   openssh
```

![yubikeys ssh setup](img/yubikey-sshkey.jpg)

### PIV-backed keys and attestations

YubiKeys are compatible with [Personal Identity Verification (PIV)] for smart
cards, which allows using them for encryption and signing operations on top
of this particular standard. Project members operating critical infrastructure
or driving releases should mind the following instructions to setup PIV-backed

> keys.

Reviewer: shouldn't everyone with a yubikey follow this? Why only "Project members operating critical infrastructure or driving releases"?
Should we just delete these 3 lines? They don't seem to add much.

Author: Agree with your suggestion! Done [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/56ee0bad39dd5568e0f06b8645ddd3766b3b6fbe)

#### Conventions and requirements

To leverage PIV capabilities, **it's mandatory to change** the PIN and PUK

> default values. You can do that using either with `ykman` or the Yubico

Reviewer: ```suggestion
default values. You can do that using either `ykman` or the Yubico
```

Authenticator app.

```shell
ykman piv access change-pin
ykman piv access change-puk
```

Although it is possible to create and manage PIV keys using the Yubico
Authenticator, you must use `ykman` to have access to security options
not available through the desktop application.

#### Generating PIV-backed keys

> When creating PIV-backed keys with your YubiKey, you must note following

Reviewer: ```suggestion
When creating PIV-backed keys with your YubiKey, you must note the following
```

security requirements:

- `ECCP256` should be favored as the encryption algorithm
- PIN confirmation must be required at least `once` when accessing a PIV slot
- human interaction against the YubiKey (touching it) is always required

To generate a key and store it into the slot `9a` (authentication use cases)
with recommended security defaults, run:

```shell
ykman piv keys generate 9a - \
  --algorithm ECCP256\

>   --pin-policy once\

Reviewer: ```suggestion
  --algorithm ECCP256 \
  --pin-policy once \
```

  --touch-policy always
```

Similarly, to generate a key and store it into the slot `9c`
(signing use cases):

```shell
ykman piv keys generate 9c - \
  --algorithm ECCP256\

>   --pin-policy once\

Reviewer: ```suggestion
  --algorithm ECCP256 \
  --pin-policy once \
```

  --touch-policy always
```

#### Exporting public keys and certificates from PIV slots

After generating keys, you can export the public keys related to the PIV
slots and store them externally:

```shell
ykman piv keys export 9a pubkey-9a.pem
ykman piv keys export 9c pubkey-9c.pem
```

These public keys are also required to generate self-signed `X.509`

> certificates and have them phisycally stored in the same PIV slots.

Reviewer: ```suggestion
certificates and have them physically stored in the same PIV slots.
```

For example, to generate and store such a certificate for the `9a` PIV slot
already storing a key:

```shell
ykman piv certificates generate 9a pubkey-9a.pem \

>   --subject CN=<your name>

Reviewer: ```suggestion
  --subject "CN=<your name>" \
```

  --valid-days 2000
```

Note that:

- `subject` should be a [RFC-4514 string]

> - `valid-days` impacts directly certificate rotation (default is 365 days) 

Reviewer: ```suggestion
- `valid-days` directly impacts certificate rotation (default is 365 days)
```

#### Exporting attestations and sharing them through team DB

In addition to self-signed certificates, you can also generate an
[attestation certificate] for a PIV-backed key pair.

```shell
ykman piv keys attest 9a attestation-9a.pem
ykman piv keys attest 9c attestation-9c.pem
```

To allow verification for these attestations, you must also export Yubico's
intermediate attestation certificate pre-loaded in the YubiKey and signed
with Yubico's root attestation CA:

```shell
ykman piv certificates export f9 f9-intermediate.pem
```

> We expect these files to be shared for further verification through the 

Reviewer: ```suggestion
We expect these files to be shared for further verification through the
```

[Project members DB]. To bring these files to `team`, you must:

- create a new folder under `team/hardware-keys`
- add to it any `attestation-9*.pem` files related to the PIV slots you set up
- add to it the `f9-intermediate.pem` attestation file for your YubiKey
- link these files with your member information under `team/people/<your-user>.toml`

Preferably, name each folder you create at `team/hardware-keys` after your YubiKey

> Serial Number. Check the [TOML schema] for addition details.

Reviewer: ```suggestion
serial number. Check the [TOML schema] for additional details.
```

app.

```shell
$ ykman fido credentials list
Enter your PIN:
Credential ID  RP ID    Username  Display name
86707903...    ssh:git  openssh   openssh
```

![yubikeys ssh setup](img/yubikey-sshkey.jpg)

### PIV-backed keys and attestations

YubiKeys are compatible with [Personal Identity Verification (PIV)] for smart
cards, which allows using them for encryption and signing operations on top
of this particular standard.

#### Conventions and requirements

Before using PIV, **you must change** the default PIN and PUK. Use `ykman` or
the Yubico Authenticator app. With `ykman`, run:

```shell
ykman piv access change-pin
ykman piv access change-puk
```

Yubico Authenticator can create and manage PIV keys, but some security
options are available only in `ykman`.

#### Generating PIV-backed keys

> When creating PIV-backed keys, follow these

Reviewer: ```suggestion
When creating PIV-backed keys, follow these
security requirements:
```
while applying my suggestion you probably deleted also this line by mistake

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-forge/1086/d7c42ceecdfbb22f47fa586ec44a1361f2af82ba..7297f285148ba40abbe51bb7ef179475f8bff6ba)*

Author: Solved [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/f6b75e396074045e9b07f3095dd46045cd54a62f)

```

#### Exporting public keys and certificates from PIV slots

After generating keys, you can export the public keys related to the PIV
slots and store them externally:

```shell
ykman piv keys export 9a pubkey-9a.pem
ykman piv keys export 9c pubkey-9c.pem
```

These public keys are also required to generate self-signed `X.509`
certificates. For example, to generate a certificate associated with
the `9a` PIV slot, run:

```shell
ykman piv certificates generate 9a pubkey-9a.pem --subject "CN=<your name>" --valid-days 2000
```

Note that:

* `subject` should be an [RFC-4514 string]
* `valid-days` sets the certificate lifetime (default is 365 days)

#### Exporting attestations and sharing them through team DB

In addition to self-signed certificates, you can also generate an
[attestation certificate] for a PIV-backed key pair.

```shell
ykman piv keys attest 9a attestation-9a.pem
ykman piv keys attest 9c attestation-9c.pem
```

To allow verification for these attestations, you must also export Yubico's
intermediate attestation certificate pre-loaded in the YubiKey and signed
with Yubico's root attestation CA:

```shell
ykman piv certificates export f9 f9-intermediate.pem
```

> Add these files to the [team DB]:

Reviewer: ```suggestion
To add these files to the [team DB], follow the following steps:
```
Sorry, my previous suggestion wasn't correct on this line.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-forge/1086/d7c42ceecdfbb22f47fa586ec44a1361f2af82ba..7297f285148ba40abbe51bb7ef179475f8bff6ba)*

Author: Fixed it [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/f6b75e396074045e9b07f3095dd46045cd54a62f)

app.

```shell
$ ykman fido credentials list
Enter your PIN:
Credential ID  RP ID    Username  Display name
86707903...    ssh:git  openssh   openssh
```

![yubikeys ssh setup](img/yubikey-sshkey.jpg)

### PIV-backed keys and attestations

YubiKeys are compatible with [Personal Identity Verification (PIV)] for smart
cards, which allows using them for encryption and signing operations on top
of this particular standard.

#### Conventions and requirements

To leverage PIV capabilities, **it's mandatory to change** the PIN and PUK
default values. You can do that using either `ykman` or the Yubico

> Authenticator app.

Reviewer: ```suggestion
Before using PIV, **you must change** the default PIN and PUK. Use `ykman` or
the Yubico Authenticator app. With `ykman`, run:
```

Author: Changed it [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7), along with other things

```shell
ykman piv access change-pin
ykman piv access change-puk
```

Although it is possible to create and manage PIV keys using the Yubico
Authenticator, you must use `ykman` to have access to security options

> not available through the desktop application.

Reviewer: ```suggestion
Yubico Authenticator can create and manage PIV keys, but some security
options are available only in `ykman`.
```

Author: Changed it [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7), along with other things

#### Generating PIV-backed keys

> When creating PIV-backed keys with your YubiKey, you must note the following

Reviewer: ```suggestion
When creating PIV-backed keys, follow these
```

Author: Changed it [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7), along with other things

security requirements:

* `ECCP256` should be favored as the encryption algorithm
* PIN confirmation must be required at least `once` when accessing a PIV slot

> * human interaction against the YubiKey (touching it) is always required

Reviewer: ```suggestion
* Prefer the `ECCP256` encryption algorithm.
* Require PIN confirmation at least `once` when accessing a PIV slot.
* Require human interaction against the YubiKey (touching it).
```

Author: See https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7

To generate a key and store it into the slot `9a` (authentication use cases)
with recommended security defaults, run:

```shell
ykman piv keys generate 9a - --algorithm ECCP256 --pin-policy once --touch-policy always
```

Similarly, to generate a key and store it into the slot `9c`
(signing use cases):

```shell
ykman piv keys generate 9c - --algorithm ECCP256 --pin-policy once --touch-policy always
```

#### Exporting public keys and certificates from PIV slots

After generating keys, you can export the public keys related to the PIV
slots and store them externally:

```shell
ykman piv keys export 9a pubkey-9a.pem
ykman piv keys export 9c pubkey-9c.pem
```

These public keys are also required to generate self-signed `X.509`
certificates and to have them physically stored in the same PIV slots.
For example, to generate and store such a certificate for the `9a` PIV slot

> already storing a key:

Reviewer: what do you mean with "already storing a key"?

Author: It seems one can store a X.509 certificate only within a PIV slot already storing a private key

```shell
ykman piv certificates generate --help
Usage: ykman piv certificates generate [OPTIONS] SLOT PUBLIC-KEY

  Generate a self-signed X.509 certificate.

  A self-signed certificate is generated and written to one of the slots on the YubiKey. A private key must already be present in the corresponding key slot.

# more stuff...
```

Happy to add a note clarifying that

Reviewer: ok, maybe if you can try to reword to add a bit of clarity, because from a first "casual"/"distracted" read I didn't get that 
"already storing a key" was referring to the "slot". 🤔

Author: Simplified it [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/7297f285148ba40abbe51bb7ef179475f8bff6ba)

```shell
ykman piv certificates generate 9a pubkey-9a.pem --subject "CN=<your name>" --valid-days 2000
```

Note that:

* `subject` should be a [RFC-4514 string]

> * `valid-days` directly impacts certificate rotation (default is 365 days)

Reviewer: ```suggestion
* `subject` should be an [RFC-4514 string]
* `valid-days` sets the certificate lifetime (default is 365 days)
```

Author: Also in https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7

#### Exporting attestations and sharing them through team DB

In addition to self-signed certificates, you can also generate an
[attestation certificate] for a PIV-backed key pair.

```shell
ykman piv keys attest 9a attestation-9a.pem
ykman piv keys attest 9c attestation-9c.pem
```

To allow verification for these attestations, you must also export Yubico's
intermediate attestation certificate pre-loaded in the YubiKey and signed
with Yubico's root attestation CA:

```shell
ykman piv certificates export f9 f9-intermediate.pem
```

We expect these files to be shared for further verification through
the [Project members DB]. To bring these files to `team`, you must:

* create a new folder under `team/hardware-keys`
* add to it any `attestation-9*.pem` files related to the PIV slots you set up
* add to it the `f9-intermediate.pem` attestation file for your YubiKey
* link these files with your member information under `team/people/<your-user>.toml`

Preferably, name each folder you create at `team/hardware-keys` after your YubiKey

> serial number. Check the [TOML schema] for additional details.

Reviewer: ```suggestion
Add these files to the [Project members DB]:

* Create a directory under [`team/hardware-keys`](https://github.com/rust-lang/team/tree/main/hardware-keys) named after your YubiKey's serial number.
* Add the `attestation-9*.pem` files for the PIV slots you configured.
* Add the YubiKey's `f9-intermediate.pem` attestation file.
* Link to these files from `team/people/<your-user>.toml`.

See the [TOML schema] for details.
```

Author: Also in https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7

[options for hardware-backed SSH key pairs]: https://developers.yubico.com/SSH
[OpenSSH built-in support for FIDO2 authentication]: https://developers.yubico.com/SSH/Securing_SSH_with_FIDO2.html
[Personal Identity Verification (PIV)]: https://developers.yubico.com/PIV
[web-based authentication flow]: https://docs.cloud.google.com/sdk/docs/authenticate#humans
[MFA method in your Google account]: https://support.google.com/accounts/answer/6103523?hl=en&co=GENIE.Platform%3DDesktop
[AWS SSO user sessions]: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sso.html
[AWS Identity Center configuration]: https://forge.rust-lang.org/infra/docs/aws-access.html
[MFA method of choice in your AWS user account]: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_mfa.html

> [Project members DB]: https://github.com/rust-lang/team

Reviewer: in the title you refer to this as "team db". Here as "project members DB". For simplicity we should pick one of those.
I prefer "team db" because it's how the repository is called.

Author: I'll change that!

Author: Changed it [here](https://github.com/rust-lang/rust-forge/pull/1086/commits/7679b29b5c518ce795c866c5a57e5ebc04845bd7), along with other things
